core/src/Kokkos_MemoryPool.hpp

```diff
@@ -233,30 +242,23 @@ class MemoryPool {
 
     const uint32_t *sb_state_ptr = sb_state_array;
 
-    s << "pool_size(" << (size_t(m_sb_count) << m_sb_size_lg2) << ")"
-      << " superblock_size(" << (1LU << m_sb_size_lg2) << ")" << std::endl;
-
-    for (int32_t i = 0; i < m_sb_count; ++i, sb_state_ptr += m_sb_state_size) {
-      if (*sb_state_ptr) {
-        const uint32_t block_count_lg2 = (*sb_state_ptr) >> state_shift;
-        const uint32_t block_size_lg2  = m_sb_size_lg2 - block_count_lg2;
-        const uint32_t block_count     = 1u << block_count_lg2;
-        const uint32_t block_used      = (*sb_state_ptr) & state_used_mask;
-
-        s << "Superblock[ " << i << " / " << m_sb_count << " ] {"
-          << " block_size(" << (1 << block_size_lg2) << ")"
-          << " block_count( " << block_used << " / " << block_count << " )"
-          << std::endl;
-      }
-    }
+    _do_print_state(s, sb_state_ptr);
 
     if (!accessible) {
       host.deallocate(sb_state_array, alloc_size);
     }
   }
 
+ private:
+  void _do_print_state(std::ostream &s, const uint32_t *sb_state_ptr) const {
+    Impl::_print_memory_pool_state(s, sb_state_ptr, m_sb_count, m_sb_size_lg2,
+                                   m_sb_state_size, state_shift,
+                                   state_used_mask);
+  }
```

Comment:
Why did you introduce `_do_print_state()`?

Reply:
so that I can move as little as possible to the implementation file. Also to avoid a million parameters in the nonmember function.

Reply:
> to avoid a million parameters in the nonmember function.

nonmember function has 7 parameters

Reply:
not sure what you meant

Reply:
yeah, if I don't separate out this part and include a bunch of things that are data members, it balloons to much more than this